cmd/object-api-options.go

```diff
@@ -225,9 +225,11 @@ func getAndValidateAttributesOpts(ctx context.Context, w http.ResponseWriter, r
 
 func parseObjectAttributes(h http.Header) (attributes map[string]struct{}) {
 	attributes = make(map[string]struct{})
-	for _, v := range strings.Split(strings.TrimSpace(h.Get(xhttp.AmzObjectAttributes)), ",") {
-		if v != "" {
-			attributes[v] = struct{}{}
+	for _, headerVal := range h.Values(xhttp.AmzObjectAttributes) {
+		for _, v := range strings.Split(strings.TrimSpace(headerVal), ",") {
```

Comment:
I think splitting the header value manually (using `strings.Split`) is not correct. That's exactly what `h.Values(...)` is doing. I think line 229 should be removed completely.

According to [RFC 7230 3.2.2](https://www.rfc-editor.org/rfc/rfc7230#section-3.2.2):
> A sender MUST NOT generate multiple header fields with the same field
> name in a message unless either the entire field value for that
> header field is defined as a comma-separated list [i.e., #(values)]
> or the header field is a well-known exception (as noted below).

> A recipient MAY combine multiple header fields with the same field
> name into one "field-name: field-value" pair, without changing the
> semantics of the message, by appending each subsequent field value to
> the combined field value in order, separated by a comma.  The order
> in which header fields with the same field name are received is
> therefore significant to the interpretation of the combined field
> value; a proxy MUST NOT change the order of these field values when

@klauspost A header should never be specified more than once in an HTTP request message, so if someone does and signature verfication fails, then that's totally fine.

Reply:
So `CanonicalHeaders` should be the comma separated, trimmed values I presume.

Reply:
`CanonicalHeaders` has the following format:
```
Lowercase(<HeaderName1>)+":"+Trim(<value>)+"\n"
Lowercase(<HeaderName2>)+":"+Trim(<value>)+"\n"
...
Lowercase(<HeaderNameN>)+":"+Trim(<value>)+"\n"
```
Most important is that headers are in lower-case and sorted. If the header contains multiple values (separated by spaces), then we should just use its trimmed value. If `<value>` contains multiple values, then the entire value must be used (trimmed). This the canonical request that AWS SDK generates for `aws s3api --debug get-object-attributes --bucket test --key xxx --object-attributes "ETag" "Checksum"`:
```
GET
/test/xxx
attributes=
host:localhost:9000
x-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date:20240821T114519Z
x-amz-object-attributes:ETag,Checksum

host;x-amz-content-sha256;x-amz-date;x-amz-object-attributes
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
2024-08-21 13:45:19,480 - MainThread - botocore.auth - DEBUG - StringToSign:
AWS4-HMAC-SHA256
20240821T114519Z
20240821/us-east-1/s3/aws4_request
26897135b1bd8e4d0bca2c3115e586fef66676802fa57162ac5ce065cc1cce43
```
It set the `x-amz-object-attributes` to `ETag,Checksum`. I don't think we should encounter any issues here. The signing method doesn't even take the multiple values into consideration. It's just a single value for the signing method. It may be different, when you specify the same header multiple times. But that's not allowed according to the HTTP spec, so I think it's fine if that doesn't work. No client should do that...

Reply:
@klauspost I checked the [source](https://github.com/minio/minio/blob/7b239ae1544cd8c8b4cd510a7cf6bbfd53f37bfb/cmd/signature-v4.go#L61-L84) and that seems to work fine. We may want to add the following test to `signature-v4_test.go`:
```go
func TestCanonicalHeaders(t *testing.T) {
	h := http.Header{
		"X-Test": []string{"value1", "value2", "value3.1,value3.2"},
	}
	got := getCanonicalHeaders(h)
	expected := "x-test:value1,value2,value3.1,value3.2\n"
	if got != expected {
		t.Errorf("Expected %s, but got %s", expected, got)
	}
}
```

Reply:
I checked and if you specify the following header: `http.Header{"X-TEST": []string{"1", "2", "3.1,3.2"}}`, then the default Go HTTP client will generate three `X-TEST` headers. Each with a single value.